Fix some pages not loading after navigating to new page and going back

zakariaridouh opened the pull request:
#### 569eca8c771831d872527d66c8108429981213b5
<pre>
Fix some pages not loading after navigating to new page and going back
<a href="https://bugs.webkit.org/show_bug.cgi?id=292001">https://bugs.webkit.org/show_bug.cgi?id=292001</a>
<a href="https://rdar.apple.com/147364853">rdar://147364853</a>

Reviewed by Brady Eidson.

After 290200@main, some pages that require us to perform a COOP process swap in response to their
Cross-Origin-Opener-Policy would not load when going back to them from another page. Example is
going to any YouTube.com page, then navigating away (for example to webkit.org) then clicking the
back button to navigate back.

Fix this by using the synchronous HistoryController API for IPC calls to the UI process in these cases.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItem const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItemShared):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::fulfillRequestWithKeyData):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::shouldGoToHistoryItem const):

Canonical link: <a href="https://commits.webkit.org/294080@main">https://commits.webkit.org/294080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1179b6803f2ff4ac06734c3c2bb4e991020bf783

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100765 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20417 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10716 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105902 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51354 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20726 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28891 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76728 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33767 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91016 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57081 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9024 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50729 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85637 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9099 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108257 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20481 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28246 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87218 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85231 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21690 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29939 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7666 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21887 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27818 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33074 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27629 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30947 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29187 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/network/soup/CertificateInfo.h:
```diff
@@ -28,7 +28,9 @@
 #include "NotImplemented.h"
+#if !PLATFORM(GTK) && !PLATFORM(WPE)
 #include <libsoup/soup.h>
+#endif
```
This isn't going to work.

Also, this is soup/CertificateInfo.h so there's no reason to not include the header. This shouldn't be in your include path if you're not building for WPE/GTK.

zakariaridouh replied:
Yeah was just trying to fix this WPE/GTK build fail caused from a new include. Probably need to gate for GTK/WPE earlier up in includes.
```
/app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/CertificateInfo.h:31:10: fatal error: libsoup/soup.h: No such file or directory
   31 | #include <libsoup/soup.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```

mcatanzaro replied:
Since I don't know what the original error was, I don't know what to suggest.

soup/CertificateInfo.h should not be in the include path except when building for WPE or GTK. If it is, the include path is messed up.

zakariaridouh replied:
[Original error](https://ews-build.webkit.org/#/builders/5/builds/105852):
```
[2202/2262] Building CXX object Tools/WebKitTestRunner/CMakeFiles/TestRunnerInjectedBundle.dir/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp.o
FAILED: Tools/WebKitTestRunner/CMakeFiles/TestRunnerInjectedBundle.dir/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp.o 
/usr/bin/ccache /usr/bin/c++ -DBUILDING_TestRunnerInjectedBundle -DBUILDING_WEBKIT=1 -DBUILDING_WITH_CMAKE=1 -DBUILDING_WPE__=1 -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DGETTEXT_PACKAGE=\"WPE\" -DHAVE_CONFIG_H=1 -DJSC_GLIB_API_ENABLED -DPAS_BMALLOC=1 -DSK_ASSUME_GL=0 -DSK_ASSUME_GL_ES=1 -DSK_ASSUME_WEBGL=0 -DSK_DISABLE_LEGACY_GL_MAKE_NATIVE_INTERFACE -DSK_DISABLE_LEGACY_IMAGE_READBUFFER -DSK_DISABLE_LEGACY_INIT_DECODERS -DSK_DISABLE_LEGACY_PNG_WRITEBUFFER -DSK_DISABLE_TRACING -DSK_DISABLE_WEBGL_INTERFACE -DSK_ENABLE_PRECOMPILE -DSK_R32_SHIFT=16 -DSK_RELEASE -DSTATICALLY_LINKED_WITH_TestRunnerShared -DSTATICALLY_LINKED_WITH_WebCoreTestSupport -DTOP_LEVEL_DIR=\"/app/webkit\" -DTestRunnerInjectedBundle_EXPORTS -DUSE_SYSTEM_EGL -D_GLIBCXX_ASSERTIONS=1 -I/app/webkit/Source -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/app/webkit/Tools/WebKitTestRunner/InjectedBundle/atspi -I/app/webkit/Tools/WebKitTestRunner/InjectedBundle/glib -I/app/webkit/Tools/WebKitTestRunner/InjectedBundle/wpe -I/app/webkit/WebKitBuild/WPE/Release -I/app/webkit/WebKitBuild/WPE/Release/WebKitTestRunner/DerivedSources/InjectedBundle -I/app/webkit/Tools/WebKitTestRunner -I/app/webkit/Tools/WebKitTestRunner/InjectedBundle -I/app/webkit/Tools/TestRunnerShared -I/app/webkit/Tools/TestRunnerShared/Bindings -I/app/webkit/Tools/TestRunnerShared/UIScriptContext -I/app/webkit/WebKitBuild/WPE/Release/JavaScriptCore/Headers -I/app/webkit/WebKitBuild/WPE/Release/JavaScriptCore/PrivateHeaders -I/app/webkit/WebKitBuild/WPE/Release/PAL/Headers -I/app/webkit/WebKitBuild/WPE/Release/WTF/Headers -I/app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders -I/app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCoreTestSupport -I/app/webkit/WebKitBuild/WPE/Release/bmalloc/Headers -I/app/webkit/WebKitBuild/WPE/Release/ANGLE/Headers -I/app/webkit/Source/ThirdParty/skia -I/app/webkit/WebKitBuild/WPE/Release/Skia/Headers/top -I/app/webkit/WebKitBuild/WPE/Release/Skia/Headers/mod -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/ForwardingHeaders -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/ForwardingHeaders/wpe -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit -isystem /usr/include/libxml2 -isystem /usr/include/harfbuzz -isystem /usr/include/avif -isystem /usr/include/wpe-1.0 -isystem /usr/include/libdrm -isystem /usr/include/at-spi2-atk/2.0 -isystem /usr/include/sysprof-6 -DASSERT_ENABLED=1 -fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-odr -Wno-stringop-overread -Wno-stringop-overflow -Wno-nonnull -Wno-array-bounds -Wno-expansion-to-defined -Wno-noexcept-type -Wno-subobject-linkage -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fasynchronous-unwind-tables -fdebug-types-section -fno-strict-aliasing -fno-exceptions -fno-rtti -fcoroutines -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=c++23 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Werror -Wno-unused-parameter -DAVIF_DLL -DWPE_ENABLE_XKB=1 -pthread -MD -MT Tools/WebKitTestRunner/CMakeFiles/TestRunnerInjectedBundle.dir/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp.o -MF Tools/WebKitTestRunner/CMakeFiles/TestRunnerInjectedBundle.dir/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp.o.d -o Tools/WebKitTestRunner/CMakeFiles/TestRunnerInjectedBundle.dir/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp.o -c /app/webkit/Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp
In file included from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/ResourceResponseBase.h:30,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/ResourceResponse.h:28,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/CrossOriginAccessControl.h:31,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/ResourceLoaderOptions.h:34,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/ResourceHandle.h:29,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/BlobResourceHandle.h:34,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/FileReaderLoader.h:33,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/Blob.h:36,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/SerializedScriptValue.h:29,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/HistoryItem.h:38,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/LocalFrameLoaderClient.h:33,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/AXCoreObject.h:33,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/AccessibilityObjectAtspi.h:23,
                 from /app/webkit/Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityNotificationHandler.cpp:35:
/app/webkit/WebKitBuild/WPE/Release/WebCore/PrivateHeaders/WebCore/CertificateInfo.h:31:10: fatal error: libsoup/soup.h: No such file or directory
   31 | #include <libsoup/soup.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```

Fixing it now.

mcatanzaro commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
```diff
@@ -135,26 +135,26 @@ class WebLocalFrameLoaderClient final : public WebCore::LocalFrameLoaderClient,
     void updateOpener(const WebCore::Frame&) final;
     void cancelPolicyCheck() final;
-    
+
```
Please remove the whitespace changes. You can land there in a separate pull request if you prefer.

beidson reviewed (approved):
Fine one you get all builds working and get the white space change cleaned up.

zakariaridouh committed:
Fix some pages not loading after navigating to new page and going back
https://bugs.webkit.org/show_bug.cgi?id=292001
rdar://147364853

Reviewed by Brady Eidson.

After 290200@main, some pages that require us to perform a COOP process swap in response to their
Cross-Origin-Opener-Policy would not load when going back to them from another page. Example is
going to any YouTube.com page, then navigating away (for example to webkit.org) then clicking the
back button to navigate back.

Fix this by using the synchronous HistoryController API for IPC calls to the UI process in these cases.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::shouldGoToHistoryItem const):
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::goToItemShared):
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::fulfillRequestWithKeyData):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldGoToHistoryItem const):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::shouldGoToHistoryItem const):

Canonical link: https://commits.webkit.org/294080@main
